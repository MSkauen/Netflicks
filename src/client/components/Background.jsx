import React, {useState} from "react";
import {useLoading} from "../lib/http/useLoading";
import {ErrorView} from "../lib/ErrorView";
import {LoadingView} from "../lib/LoadingView";
import missing from "../../shared/img/missing.png";

function Background ({backgroundUrl, isRow}) {
    const [data] = useState(backgroundUrl)
    const filteredBackgrounds = data.find(res => res.type === "background");
    const filteredPosters = data.find(res => res.type === "poster");

    if(data.length !== 0) {
        if(isRow && filteredBackgrounds && !data.length < 1) {
            return <img key={data.id + 100} className="movie_poster_landscape" src={filteredBackgrounds.resolutions.original.url} alt={data.name}/>
        } else if (!filteredBackgrounds) {
            return <img key={data.id + 100} className="movie_poster_landscape" src={filteredPosters.resolutions.original.url} alt={data.name}/>
        } else if (data){
            return <img key={data.id + 100} className="movie_banner" src={filteredBackgrounds.resolutions.original.url} alt={data.name}/>
        }
    } else {
        return <img className="movie_poster_landscape" src={missing} alt=""/>
    }
}


export default function getBackground ({movieApi, movieId, isRow}) {

    const { data: data, loading, error, reload } = useLoading(
        async () => await movieApi.getMovieBackground(movieId),
        []
    );

    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !data) {
        return <LoadingView />;
    }

    return <Background backgroundUrl={data} isRow={isRow}/>;
}
