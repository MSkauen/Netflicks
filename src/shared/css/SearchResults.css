.search_results_row {
    margin-left: 30px;
    color: #fff;
    height: 100%;
}
.search_results_row .movie_poster {
    flex-wrap: wrap;
}

.search_results_posters {
    padding: 20px;
    display: flex;
    flex-wrap: wrap;
}

.search_results_row .movie {
    width: auto;
}