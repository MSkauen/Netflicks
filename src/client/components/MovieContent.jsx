import {useLoading} from "../lib/http/useLoading";
import {ErrorView} from "../lib/ErrorView";
import {LoadingView} from "../lib/LoadingView";
import React, {useState} from "react";
import Background from "./Background";

function MovieContent ({ data, movieApi }) {
    const [movie] = useState(data);

    return (
        <div className="row">

            <h2 className="search_results_title">Episodes</h2>

            <select className="banner_button" onChange={handleSeasonChange} value={sortGenre}>
                <option value="">Season 1</option>
                <option value="Action">Action</option>
                <option value="Comedy">Comedy</option>
                <option value="Crime">Crime</option>
                <option value="Drama">Drama</option>
                <option value="Horror">Horror</option>
                <option value="Romance">Romance</option>
                <option value="Science-Fiction">Science-fiction</option>
                <option value="Thriller">Thriller</option>
            </select>



            <h2>{title}</h2>
            <div className="movie_posters">

                    <a key={movie.id} href={`/show/${movie.id}`} className="movie">
                            <>
                                <div className="movie_details">
                                    <h4>{movie.name}</h4>
                                    <span>{movie.rating.average}</span>
                                </div>

                                <Background movieApi={movieApi} movieId={movie.id} isRow/>
                            </>
                    </a>

            </div>
        </div>
    )
}

export default function GetMovieRow ({movieApi, title, genre, isLargeRow}) {
    const { data: data, loading, error, reload } = useLoading(
        async () => await movieApi.getAllMovies(),
        []
    );

    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !data) {
        return <LoadingView />;
    }

    if(data && genre) {
        const filteredMovies = data.filter(movie => movie.genres.includes(genre));
        return <MovieRow data={filteredMovies.splice(0, 15)} title={title} isLargeRow={isLargeRow} />;
    }

    return <MovieRow movieApi={movieApi} data={data.splice(0, 15)} title={title} isLargeRow={isLargeRow}/>;
}
