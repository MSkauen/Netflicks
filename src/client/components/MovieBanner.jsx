import React, { useState } from "react";
import Background from "./Background";
import { ErrorView } from "../lib/ErrorView";
import truncate from "../lib/truncate";
import { LoadingView } from "../lib/LoadingView";
import { useLoading } from "../lib/http/useLoading";
import "../../shared/css/MovieBanner.css";
import {useParams} from "react-router";

function MovieBanner ({movieApi, data}) {
    const [movie] = useState(data)
    const cleanMovieSummary = movie.summary.replace(/<\/?[^>]+(>|$)/g, "");
    console.log(data)
    return (
        <header className="banner">
            <div className="banner_contents">
                <Background movieApi={movieApi} movieId={movie.id}/>
                <h1 className="banner_title">{data.name}</h1>
                <div className="banner_buttons">
                    <button className="banner_button">Play</button>
                    <button className="banner_button">Add to My list</button>
                </div>
                <h1 className="banner_description">
                    {truncate(cleanMovieSummary, 150)}
                </h1>
            </div>

            <div className="banner_fadeBottom"/>
        </header>
    )
}

export default function GetRandomMovieBanner ({movieApi}) {
    const { data: data, loading, error, reload } = useLoading(
        async () => await movieApi.getRandomMovie(),
        []
    );

    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !data) {
        return <LoadingView />;
    }

    return <MovieBanner movieApi={movieApi} data={data}/>;
}

export function GetSingleMovieBanner ({movieApi}) {
    const { id } = useParams();

    const { data: data, loading, error, reload } = useLoading(
        async () => await movieApi.getMovieSummary(id),
        []
    );

    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !data) {
        return <LoadingView />;
    }

    return <MovieBanner movieApi={movieApi} data={data}/>;
}