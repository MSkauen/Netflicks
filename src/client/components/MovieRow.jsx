import React, {useState} from "react";
import {ErrorView} from "../lib/ErrorView";
import {LoadingView} from "../lib/LoadingView";
import {useLoading} from "../lib/http/useLoading";
import "../../shared/css/MovieRow.css";
import Background from "./Background";

function MovieRow ({movieApi, data, title, isLargeRow}) {
    const [movies] = useState(data);

    return (
        <div className="row">
            <h2>{title}</h2>
                <div className="movie_posters">
                    {movies.map(movie => (
                        <a key={movie.id} href={`/show/${movie.id}`} className="movie">
                            {!isLargeRow ? (
                                <img className={`movie_poster ${isLargeRow && "movie_poster_landscape"}`}
                                     src={movie.image.original} alt={movie.name}/>
                            ) : (
                                <>
                                    <div className="movie_details">
                                        <h4>{movie.name}</h4>
                                        <span>{movie.rating.average}</span>
                                    </div>
                                    <Background movieApi={movieApi} movieId={movie.id} isRow/>
                                </>
                                )}
                        </a>
                    ))}
                </div>
            </div>
    )
}

export default function GetMovieRow ({movieApi, title, genre, isLargeRow}) {
    const { data: data, loading, error, reload } = useLoading(
        async () => await movieApi.getAllMovies(),
        []
    );

    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !data) {
        return <LoadingView />;
    }

    if(data && genre) {
        const filteredMovies = data.filter(movie => movie.genres.includes(genre));
        return <MovieRow data={filteredMovies.splice(0, 15)} title={title} isLargeRow={isLargeRow} />;
    }

    return <MovieRow movieApi={movieApi} data={data.splice(0, 15)} title={title} isLargeRow={isLargeRow}/>;
}

