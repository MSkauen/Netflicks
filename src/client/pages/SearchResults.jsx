import React from "react";
import GetMovieBanner from "../MovieBanner";
import Navbar from "../Navbar";
import "../../shared/css/App.css";

export function SearchResults ({movieApi}) {
    return <div className="App">
        <Navbar movieApi={movieApi}/>
        <GetMovieBanner movieApi={movieApi}/>
    </div>
}