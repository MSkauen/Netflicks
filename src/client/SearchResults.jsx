import React, {useEffect, useState} from "react";
import {ErrorView} from "./lib/ErrorView";
import {LoadingView} from "./lib/LoadingView";
import {useLoading} from "./lib/http/useLoading";
import "../shared/css/MovieRow.css";
import "../shared/css/SearchResults.css";
import Background from "./Background";
import {useParams} from "react-router";

function SearchResults ({movieApi, data, title}) {
    const [movies] = useState(data);
    const [sortType, setSortType] = useState(false);
    let [sortedMovies] = useState(movies);

    const sortedMoviesAsc = [].concat(movies)
        .sort((a, b) => a.show.rating.average > b.show.rating.average ? 1 : -1);

    const sortedMoviesDes = [].concat(movies)
        .sort((a, b) => a.show.rating.average < b.show.rating.average ? 1 : -1);

    const sort = () => {
        setSortType(current => !current)
    };

    if (sortType === false) {
        sortedMovies = sortedMoviesDes;
    } else if (sortType === true) {
        sortedMovies = sortedMoviesAsc;
    }


    return (
        <div className="search_results_row">
            <h2>Search results for '{title}': {data.length} titles</h2>
            <button className="banner_button" onClick={sort}>Sort by rating</button>
            <div className="search_results_posters">
                {sortedMovies.map(movie => (
                    <div className="movie">
                        <div className="movie_details">
                            <h4 key={movie.show.name}>{movie.show.name}</h4>
                            <span key={"r" + movie.id}>{movie.show.rating.average}</span>
                        </div>
                        <Background key={movie.show.id} movieApi={movieApi} movieId={movie.show.id} isRow/>
                    </div>
                ))}
            </div>
        </div>
    )
}

export default function GetSearchResults ({movieApi, title}) {
    const { id } = useParams()

    const { data: data, loading, error, reload } = useLoading(
        async () => await movieApi.getMovie(id),
        []
    );

    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !data) {
        return <LoadingView />;
    }

    return <SearchResults movieApi={movieApi} data={data.splice(0, 15)} title={id}/>;
}

