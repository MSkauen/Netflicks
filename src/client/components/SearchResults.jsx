import React, { useState} from "react";
import {ErrorView} from "../lib/ErrorView";
import {LoadingView} from "../lib/LoadingView";
import {useLoading} from "../lib/http/useLoading";
import "../../shared/css/MovieRow.css";
import "../../shared/css/SearchResults.css";
import Background from "./Background";
import {useParams} from "react-router";

export function SearchResults ({movieApi, data, query}) {
    const [movies] = useState(data);
    const [sortType, setSortType] = useState(false);
    const [sortGenre, setSortGenre] = useState('');

    let [sortedMovies] = useState(movies);

    const sortedMoviesAsc = [].concat(movies)
        .sort((a, b) => a.show.rating.average > b.show.rating.average ? 1 : -1);

    const sortedMoviesDes = [].concat(movies)
        .sort((a, b) => a.show.rating.average < b.show.rating.average ? 1 : -1);
    const sort = () => {
        setSortType(current => !current)

    };
    const handleCategoryChange = (e) => {
        setSortGenre(e.target.value);

    };

    //This is not great
    if (sortType === false) {
        sortedMovies = sortedMoviesDes;
    } else if (sortType === true) {
        sortedMovies = sortedMoviesAsc;
    }
    if(data && sortGenre) {
        sortedMovies = sortedMovies.filter(movie => movie.show.genres.includes(sortGenre));
    }

    return (
        <div className="search_results_row">
            <h2 className="search_results_title">Search results for '{query}': {data.length} titles</h2>
            <button className="banner_button" onClick={sort}>Sort by rating</button>

            <select className="banner_button" onChange={handleCategoryChange} value={sortGenre}>
                <option value="">Sort by category</option>
                <option value="Action">Action</option>
                <option value="Comedy">Comedy</option>
                <option value="Crime">Crime</option>
                <option value="Drama">Drama</option>
                <option value="Horror">Horror</option>
                <option value="Romance">Romance</option>
                <option value="Science-Fiction">Science-fiction</option>
                <option value="Thriller">Thriller</option>
            </select>

            <div className="search_results_posters">
                {sortedMovies.map(movie => (
                    <a key={movie.show.id} href={`/show/${movie.show.id}`} className="movie">
                        <div className="movie_details">
                            <h4>{movie.show.name}</h4>
                            <span>{movie.show.rating.average}</span>
                        </div>
                        <Background movieApi={movieApi} movieId={movie.show.id} isRow/>
                    </a>
                ))}
            </div>
        </div>
    )
}

export default function GetSearchResults ({movieApi}) {
    const { id } = useParams()

    const { data: data, loading, error, reload } = useLoading(
        async () => await movieApi.getMovie(id),
        [id]
    );

    if (error) {
        return <ErrorView error={error} reload={reload()} />;
    }

    if (loading || !data) {
        return <LoadingView />;
    }

    return <SearchResults movieApi={movieApi} data={data.splice(0, 15)} query={id}/>;
}

