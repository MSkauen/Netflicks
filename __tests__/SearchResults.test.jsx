import React from "react";
import ReactDOM from "react-dom";
import { describe, expect, it } from "@jest/globals";
import { act } from "react-dom/test-utils";
import {SearchResults} from "../src/client/components/SearchResults";
import {MemoryRouter} from "react-router";

async function render(component) {
    const container = document.createElement("div");
    document.body.appendChild(container);
    await act(async () => {
        ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
    });
    return container;
}

describe("search results", () => {
    it("show search results on dom", async () => {
        const movieApi = {
            getMovieBackground: async () => [
                {
                    id: 1,
                    resolutions: {
                        original: {
                            url: "https://static.tvmaze.com/uploads/images/medium_portrait/0/1.jpg"
                        }
                    },
                    type: "background"
                },
            ]
        };
        const container = await render(<SearchResults movieApi={ movieApi } data={[]} query={"Dome"} />);

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h2").textContent).toContain("Dome");
    });
})
