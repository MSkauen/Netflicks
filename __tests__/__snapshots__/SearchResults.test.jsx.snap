// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search results show search results on dom 1`] = `"<div class="search_results_row"><h2 class="search_results_title">Search results for 'Dome': 0 titles</h2><button class="banner_button">Sort by rating</button><select class="banner_button"><option value="">Sort by category</option><option value="Action">Action</option><option value="Comedy">Comedy</option><option value="Crime">Crime</option><option value="Drama">Drama</option><option value="Horror">Horror</option><option value="Romance">Romance</option><option value="Science-Fiction">Science-fiction</option><option value="Thriller">Thriller</option></select><div class="search_results_posters"></div></div>"`;
